README.md
# TSP
[Traveling Salesman Problem]
Bonus project for 2018 NTHU CS Data Structure Course. The project is to solve 2 variations of classical TSP problem. 

In the first variation, each city has offered happiness point to the salseman. However, happiness point of each city can only be earned once. Also, the salse man has limited traveling time to explore cities. This means that the salseman need not explore all the cities. The goal is to earn maximum happiness point within the limited time. The route has to be start and end at the same city.

The second variation inherit the first one. However, each city offer its happiness point only during the period of open and close time. The salseman can spend time staying at the city until its open time to earn the happiness point.

To compile the project:
    g++ -std=c++11 -O3 main.cpp Tsp.cpp -o m.exe

Exection the project:
    ./m.exe [foldername]
    (The folder has to contain the map file: tp.data)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.0.1 - 2018-12-09
### Added
- A simple greedy solution for the sub1 problem.

Tsp.cpp
#include <chrono>
#include <iostream>
#include <fstream>
#include <sstream>
#include <map>
#include "Tsp.hpp"
using namespace std;

Tsp::~Tsp() {};
bool Tsp::hasError() {return errorFlag;}
void Tsp::read(int &argc, char* argv[]) {
    constexpr auto &&now = std::chrono::high_resolution_clock::now;
    auto start = now();

    if (argc < 2) {
        errorFlag = true;
        cout <<  "Missing command argument. Please specify the path to tp.data." << endl;
        return;
    }

    filePath = string(argv[1]);
    ifstream in(filePath + "/tp.data");
    if (!in.is_open()) {
        errorFlag = true;
        cout << "Unable to open \"" 
            << filePath << "/tp.data\". Please check existance of file or path." << endl;
        return;
    }

    string _tmp;
    getline(in, _tmp);
    stringstream ss(_tmp);
    ss >> nodes >> edgeNum >> totalTimeQuota >> startTime;
    cities.reserve(nodes);
    edges.reserve(edgeNum);
    
    cout << "Reading Cities...";
    map<string, int> cityNameAndIndex;
    short count = nodes;
    while(count--) {
        getline(in, _tmp);
        ss.clear(); ss.str(_tmp);
        string name;
        int hapiness, start, close;
        ss >> name >> hapiness >> start >> close;
        int index = cities.size();
        cities.emplace_back(index, name, hapiness, start, close);
        cityNameAndIndex[name] = index;

    }
    cout << "complete." << endl;
    
    
    shortestTravelTime.resize(nodes);
    cameFrom.resize(nodes);
    for (int i = 0 ; i < nodes ; ++i) {
        shortestTravelTime[i].resize(nodes);
        cameFrom[i].resize(nodes);
        for (int j = 0 ; j < nodes ; ++j) {
            if (i == j) shortestTravelTime[i][j] = 0;
            else shortestTravelTime[i][j] = 2147483647;
            cameFrom[i][j] = -1;
        }
    }
    
    cout << "Reading Edges...";
    count = edgeNum;
    while(count--) {
        getline(in, _tmp);
        ss.clear(); ss.str(_tmp);
        string name1, name2;
        int travelTime;
        ss >> name1 >> name2 >> travelTime;

        int index_city1 = cityNameAndIndex.find(name1)->second;
        City &c1 = cities[index_city1];

        int index_city2 = cityNameAndIndex.find(name2)->second;
        City &c2 = cities[index_city2];
        edges.emplace_back(&c1, &c2, travelTime);

        shortestTravelTime[c1.index][c2.index] = shortestTravelTime[c2.index][c1.index] = travelTime;
        cameFrom[c1.index][c2.index] = c1.index;
        cameFrom[c2.index][c1.index] = c2.index;
    }
    cout << "complete." << endl;

    chrono::duration<double> elapsed = now() - start;
    cout << "readData, completed in : " << (int) elapsed.count() * 1000 << "ms." << endl;

    // printCities();
    // printEdges();
    // printShortestTravelTime();
    // printCamefrom();

    start = now();
    calcWarshall();
    elapsed = now() - start;
    cout << "calculate Warshall, completed in : " << (int) elapsed.count() * 1000 << "ms." << endl;
};

void Tsp::openTime(bool b) {
    takeOpentime = b;
};

// void Tsp::recordPath(Record &record, City* city, int arriving, int leaving) {
//     record.path.emplace_back(city->name, arriving, leaving);
// };

void Tsp::sub1_greedy() {
    if (!reseted) reset();

    // set current position to starting city
    City* start = &cities[0];
    City* current = &cities[0];
    int totalEarned = 0, timeQuota = totalTimeQuota, currentTime = startTime;
    map<float, City*> cpRatio; // highest cp element is at the end
    vector<City*> unvisited; unvisited.reserve(cities.size());
    for (int i = 0 ; i < cities.size() ; ++i) {
        unvisited.push_back(&cities[i]);
    }

    // record start position;
    start->visited = true;
    totalEarned += start->hapiness;
    greedyResult.path.emplace_back(start->name, currentTime, currentTime);
    
    // 1. while loop:
    int loopCount = 0;
    const int limit = nodes + 3;
    while(true) {

        // 1b. remove destination that do not fit to remaining time quota
        greedy_filterUnvisited(unvisited, start, current, timeQuota);
        if (unvisited.empty()) break;

        // 1c. calculate c/p ratio of earnable hapiness vs travel time for a not visited city
        cpRatio.clear();
        greedy_calcCP(unvisited, cpRatio, current);

        // 1e. take highest cp city as destination, go to that city and record path
        auto it = cpRatio.rbegin();
        City* dest = it->second;


        greedy_goto(current, dest, currentTime, totalEarned, timeQuota);
        current = dest;

        if (loopCount++ > limit) {
            string message = string("Tsp::greedySolution, while loop exceed limit times(")
                + to_string(limit) + ").";
            throw runtime_error(message);
        }
    }

    // 2. go back to starting point.
    greedy_goto(current, start, currentTime, totalEarned, timeQuota);

    // 3. save greedy result
    greedyResult.hapiness = totalEarned;
    greedyResult.time = greedyResult.path.back().leaving - startTime;
};

void Tsp::greedy_filterUnvisited(vector<City*> &unvisited, City* start, City* current, int timeQuota) {
    for (int i = unvisited.size() - 1 ; i >= 0 ; --i) {
        City* dest = unvisited[i];
        bool erase = false;
        if (dest->visited) {
            erase = true;
        } else {
            int requiredTime = shortestTravelTime[current->index][dest->index]
                + shortestTravelTime[dest->index][start->index];
            if (timeQuota < requiredTime) erase = true;
        }
        if (erase) {
            unvisited.erase(unvisited.begin() + i);
        }
    }
};

void Tsp::greedy_calcCP(vector<City*> &unvisited, map<float, City*> &cpRatio, City* source) {
    for (int i = 0 ; i < unvisited.size(); ++i) {
        int earnable = 0;
        City* dest = unvisited[i];
        City* current = unvisited[i];

        int loops = 0;
        const int limit = nodes + 3;
        while(current != source) {

            if (!current->visited) earnable += current->hapiness;

            // move current to cameFrom
            current = &cities[cameFrom[source->index][current->index]];

            if (loops++ > limit) {
                string message = string("Tsp::greedy_calcCP while loop exceeds limit times(")
                    + to_string(limit) + ")";
                throw runtime_error(message);
            };
        }

        float cp = earnable / shortestTravelTime[source->index][dest->index];
        cpRatio[cp] = dest;
    }
};

void Tsp::greedy_goto(City* source, City* destination,
        int &currentTime, int &totalEarned, int &timeQuota) {
    vector<City*> passingCities;
    City* current = destination;

    int loops = 0;
    const int limit = nodes + 3;
    while(current != source) {

        if (!current->visited) {
            current->visited = true;
            totalEarned += current->hapiness;
        }

        // move current to cameFrom
        passingCities.push_back(current);
        current = &cities[cameFrom[source->index][current->index]];

        if (loops++ > limit) {
            string message = string("Tsp::greedy_goto while loop exceeds limit times: "
                + to_string(limit));
            throw runtime_error(message);
        }
    }

    // current is source now
    for (int i = passingCities.size() - 1 ; i >= 0 ; --i) {
        // record next city
        City* next = passingCities[i];
        currentTime += shortestTravelTime[current->index][next->index];
        greedyResult.path.emplace_back(next->name, currentTime, currentTime);
        current = next;
    }
    timeQuota -= shortestTravelTime[source->index][destination->index];
};

void Tsp::findSolution() {
    sub1_greedy();
    saveFileSub1(greedyResult);
    showVisitStatus();
};

void Tsp::reset() {
    reseted = true;
    for (int i = 0; i < cities.size(); ++i) {
        cities[i].visited = false;
    }
};

void Tsp::saveFileSub1(Record &record) {
    ofstream o(filePath + "/ans1.txt");
    stringstream ss;
    ss << record.hapiness << ' ' << record.time << endl;
    for (int i = 0 ; i < record.path.size() ; ++i) {
        FootPrint &f = record.path[i];
        ss << f.name << ' ' << f.arriving << ' ' << f.leaving << endl;
    }
    o << ss.rdbuf();
}

void Tsp::calcWarshall() {
    cout << "Start to calculate Warshall" << endl;
    for (short k = 0 ; k < nodes ; ++k) {
        for (short i = 0 ; i < nodes ; ++i) {
            for (short j = 0 ; j < nodes; ++j) {
                if (shortestTravelTime[i][j] > shortestTravelTime[i][k] + shortestTravelTime[k][j]) {
                    shortestTravelTime[i][j] = shortestTravelTime[i][k] + shortestTravelTime[k][j];
                    cameFrom[i][j] = cameFrom[k][j];
                }
            }
        }
    }
    cout << "Warshall calculate complete.";
    // saveShortestTravelTime();
    // saveCameFrom();
};

void Tsp::printData() {
    cout << "Cities: " << endl;
    for (int i = 0 ; i < nodes ; ++i) {
        City &c = cities[i];
        if (i + 1 < 10) cout << "  ";
        else if (i + 1 < 100) cout << ' ';

        cout << i+1 << ". " << c.name;

        if (c.name.length() == 2) cout << "  ";
        else if (c.name.length() == 3) cout <<  ' ';

        cout << "  hapiness: ";
        if (c.hapiness < 10) cout << "  ";
        else if (c.hapiness < 100) cout << " ";
        
        cout << c.hapiness;
        cout << endl;
    }

    cout << "Edges:" << endl;
    for (int i = 0 ; i < edgeNum ; ++i) {
        Edge &e = edges[i];
        cout << "  " << i+1 << ". " << e.city1->name << "  " << e.city2->name << "  " << e.travelTime << endl;
    }
};

void Tsp::printCities() {
    for (int i = 0 ; i < cities.size() ; ++i) {
        City &c = cities[i];
        cout << c.index << ". " << c.name << ", " << c.hapiness << endl;
    }
    cout << endl;
};


void Tsp::printEdges() {
    for (int i = 0 ; i < edges.size() ; ++i) {
        Edge &e = edges[i];
        cout << i << ". " << e.city1->name << " " << e.city2->name << ", " << e.travelTime << endl;
    }
    cout << endl;
};

void Tsp::printShortestTravelTime() {
    cout << "Printing shortestTravelTime..." << endl;
    for (short i = 0 ; i < nodes ; ++i) {
        for (short j = 0 ; j < nodes ; ++j) {
            cout << shortestTravelTime[i][j];
            if (j < nodes - 1) cout << ' ';
            else cout << endl;
        }
    }
    cout << endl;
};

void Tsp::printCamefrom() {
    cout << "Print cameFrom...." << endl;
    for (short i = 0 ; i < nodes ; ++i) {
        for (short j = 0 ; j < nodes ; ++j) {
            cout << cameFrom[i][j];
            if (j < nodes - 1) cout << ' ';
            else cout << endl;
        }
    }
    cout << endl;
};

void Tsp::saveShortestTravelTime() {
    ofstream o(filePath + "/shortestTravelTime.txt");
    for (short i = 0 ; i < nodes ; ++i) {
        for (short j = 0 ; j < nodes ; ++j) {
            o << shortestTravelTime[i][j];
            if (j < nodes - 1) o << ' ';
            else o << endl;
        }
    }
    cout << filePath << "/shortestTravelTime.txt saved." << endl;
};
void Tsp::saveCameFrom() {
    ofstream o(filePath + "/cameFrom.txt");
    for (short i = 0 ; i < nodes ; ++i) {
        for (short j = 0 ; j < nodes ; ++j) {
            o << cameFrom[i][j];
            if (j < nodes - 1) o << ' ';
            else o << endl;
        }
    }
    cout << filePath << "/cameFrom.txt saved." << endl;
};

void Tsp::showVisitStatus() {
    int visited = 0 , unvisited = 0;
    for (int i = 0 ; i < cities.size() ; ++i) {
        if (cities[i].visited) ++visited;
        else ++unvisited;
    };
    cout << "Total City: " << cities.size() << ", visited/unvisited: "
         << visited << "/" << unvisited << endl;
}

Tsp.hpp
/* This code is designed to calculate 2 kinds of variation TSP problem
    with no more than 100 citys
    with execution time less than 30 sec.
*/
#pragma once
#ifndef _TSP_YUNJING_
#define _TSP_YUNJING_
#include <vector>
#include <map>
using namespace std;

class City {
public:
    City() {};
    City(short _i, string _n, int _hapiness, int _s = 0, int _c = 1400)
        :index(_i), name(_n), hapiness(_hapiness), start(_s), close(_c) {};
    string name;
    short index;
    int hapiness = 0,
        start = 0, close = 1440;
    bool visited = false;
};

class Edge {
public:
    Edge() {};
    Edge(City* _c1, City* _c2, int _t)
        : city1(_c1), city2(_c2), travelTime(_t) {};
    City* city1;
    City* city2;
    int travelTime;
};
class FootPrint {
public:
    FootPrint() {};
    FootPrint(string _name, int _arriving, int _leaving)
        :name(_name), arriving(_arriving), leaving(_leaving) {};
    string name;
    int arriving, leaving;
};
class Record {
public:
    int hapiness, time;
    vector<FootPrint> path;
};

class Tsp {
public:
    Tsp() {};
    ~Tsp();
    bool hasError();
    void openTime(bool b);
    void read(int &argc, char* argv[]);
    void simulatedAnnealing();
    void findSolution();
    // void outputResult();

    /* for test */
    void printData();
    void saveShortestTravelTime();
    void saveCameFrom();

private:
    bool errorFlag = false;
    bool takeOpentime = true;
    bool reseted = true;

    short chooseCity = 4;

    int nodes = 0, edgeNum = 0,
        totalTimeQuota = 0, startTime = 480;
    vector<vector<int>> shortestTravelTime, cameFrom;

    string filePath;
    vector<City> cities;
    vector<Edge> edges;
    Record greedyResult;
    
    void reset(); // reset data for a new round of algorithm.
    void saveFileSub1(Record &record);
    void calcWarshall(); // use warshall to calculate all shortestTravelTime
    // void recordPath(Record &record, City* city, int arriving, int leaving);

    void sub1_greedy();
    void greedy_filterUnvisited(vector<City*> &unvisited, City* start, City* current, int timeQuota);
    void greedy_calcCP(vector<City*> &unvisited, map<float, City*> &cpRatio, City* source);
    void greedy_goto(City* source, City* destination, 
        int &currentTime, int &totalEarned, int &timeQuota);

    /* for test*/
    void printCities();
    void printEdges();
    void printShortestTravelTime();
    void printCamefrom();
    void showVisitStatus();
};
#endif /* _TSP_YUNJING_ */

TestCaseGenerate.cpp
#include <iostream>
#include <stdexcept>
#include <fstream>
#include <sstream>
#include <string>
#include <chrono>
#include <random>
#include <set>
using namespace std;

class coordinate {
public:
    coordinate() {};
    coordinate(double _i, double _j) : i(_i), j(_j) {};
    double i, j;
};
class coordCompare {
public:
    bool operator() (const coordinate& lhs, const coordinate& rhs) const {
        if (lhs.i != rhs.i) return lhs.i < rhs.i;
        return lhs.j < rhs.j;
    }
};
class City {
public:
    City() {};
    City(int _index, string _n, double x, double y, short _hapiness = 1,
            int _l = 4, int _s = 0, int _c = 1400)
        :index(_index), name(_n), hapiness(_hapiness), start(_s), close(_c),
        level(_l), coord(x, y) {};
    int index;
    string name;
    short hapiness;
    int start;
    int close;
    short level;
    coordinate coord;
};
class Edge {
public:
    Edge() {};
    Edge(City* _c1, City* _c2, int _level, int _distance, int _travelTime)
        : city1(_c1), city2(_c2), level(_level), distance(_distance), travelTime(_travelTime) {};
    City* city1;
    City* city2;
    int level;
    double distance;
    int travelTime;
};

class Generator {
public:
    Generator(int _n);
    void addCity(int level, int x, int y);
    void addEdge(int index1, int index2, int edgeLevel);
    void generate();
    void saveFile();
    void printTest();
    void printStatus();
private:
    int totalCities, totalEdges = 0,
        timeQuota = 4*24*60, startTime = 480;
    int currentCities = 0, currentEdges = 0;
    vector<City> cities;
    vector<Edge> edges;
    vector<string> cityNames;
    vector<coordinate> coords;

    Generator() {};
    mt19937 randomGenerator;
    uniform_int_distribution<short> unidist2To4, unidistChar,
        unidistHappy, uniTravelTime, unidist0To100, unidist1To100;
    void generateCity(int level, int count);
    void generateEdgeByCoords();
    short getHapiness(short level);
    void getStartAndclose(short level, int &start, int &close);
    int getTravelTime(short level, double distance);
    bool isDuplicatedEdge(int index1, int index2);
};

void addCityAndEdge(Generator &g);

int main(int argc, char* argv[]) {

    if (argc < 2) {
        cout << "Missing command parameter." << endl
            << "Please specify mode(coord/random) and the Node number.(2-100)" << endl;
        return 1;
    }

    string arg = argv[1];
    int nodes;
    bool convertError = false;
    try {
        std::size_t pos;
        nodes = std::stoi(arg, &pos);
        if (pos < arg.size()) {
            std::cerr << "Trailing characters after number: " << arg << '\n';
        }
    } catch (std::invalid_argument const &ex) {
        std::cerr << "Invalid number: " << arg << '\n';
        convertError = true;
    } catch (std::out_of_range const &ex) {
        std::cerr << "Number out of range: " << arg << '\n';
        convertError = true;
    }

    if (convertError) {
        cout << "The input argurment is not a valid number!" << endl;
        return 1;
    }
    if (nodes > 100) {
        cout << "The input nodes cannot be greater than 100." << endl;
        return 1;
    }
    if (nodes < 2) {
        cout << "At least two node is needed!" << endl;
    }
    cout << "** nodeNumbers : " << nodes << endl;

    Generator g(nodes);
    addCityAndEdge(g);

    cout << "** Generating map..." << endl;
    g.generate();
    cout << "** Generating map complete." << endl;
    g.printStatus();

    cout << "** Saving file..." << endl;
    string filePath = "testCase/" + to_string(nodes) + "/tp.data";
    g.saveFile();
    cout << "** File Saved. Program finished. Bye ~" << endl;

    // g.printTest();
    return 0;
};

Generator::Generator(int _n) : totalCities(_n),
        unidist2To4(2, 4), unidistChar(65, 90), unidistHappy(5, 25),
        uniTravelTime(3, 30), unidist0To100(0, 100), unidist1To100(1, 100),
        randomGenerator(chrono::high_resolution_clock::now().time_since_epoch().count()) {
    cities.reserve(totalCities);
    edges.reserve(totalCities * (totalCities-1) / 2);
    //edgeNum = nodes * (nodes - 1) / 2;
};

void Generator::addCity(int level, int x, int y) {
    int nameLength = unidist2To4(randomGenerator);
    string name;

    // generate a non-repetitive valid name
    while (true) {
        for (int k = 0 ; k < nameLength ; ++k) {
            name.push_back((char) unidistChar(randomGenerator));
        }
        bool found = false;
        for (int i = 0 ; i < cityNames.size() ; ++i) {
            if (cityNames[i] == name) {
                found = true;
                break;
            }
        }
        if (found) {
            name.clear();
        } else {
            break;
        }
    }

    coords.emplace_back(x, y);
    int hapiness = getHapiness(level);
    int start, close;
    getStartAndclose(level, start, close);
    int index = cities.size();
    cities.emplace_back(index, name, x, y, hapiness, level, start, close);
    cityNames.push_back(name);
    ++currentCities;
};
void Generator::addEdge(int index1, int index2, int edgeLevel) {
    // detect duplicated
    City &c1 = cities[index1];
    City &c2 = cities[index2];
    // cout << "adding Edge: "
    //     << "city1:cities["<<index1<<"]("<<cities[index1].name<<"), "
    //     << "city2:cities["<<index2<<"]("<<cities[index2].name<<"). " << endl;
    
    if (isDuplicatedEdge(index1, index2)) return;

    double distance = sqrt((c1.coord.i - c2.coord.i) * (c1.coord.i - c2.coord.i)
            + (c1.coord.j - c2.coord.j) * (c1.coord.j - c2.coord.j));
    int travelTime = getTravelTime(edgeLevel , distance);

    // cout << "Adding edge index: " << edges.size();
    edges.emplace_back(&c1, &c2, edgeLevel, distance, travelTime);
    Edge &e = edges[edges.size() -1];
    
    // cout << ", cities: " << e.city1->name << ", " << e.city2->name 
    //     << ", edge level: " << e.level << ", travelTime: " << e.travelTime << endl;
    ++currentEdges;
};
void Generator::generate() {
    cout << "generateCity(3, 20);" << endl;
    generateCity(3, 20);
    generateCity(4, totalCities - cities.size());
    cout << "generateEdgeByCoords();" << endl;
    generateEdgeByCoords();
};

void Generator::generateCity(int level, int count) {
    for (int i = 0 ; i < count ; ++i) {
        int nameLength = unidist2To4(randomGenerator);
        string name;

        // generate a non-repetitive valid name
        while (true) {
            for (int k = 0 ; k < nameLength ; ++k) {
                name.push_back((char) unidistChar(randomGenerator));
            }
            bool found = false;
            for (int v = 0 ; v < cityNames.size() ; ++v) {
                if (cityNames[v] == name) {
                    found = true;
                    break;
                }
            }
            if (found) {
                name.clear();
                continue;
            } else {
                break;
            }
        }
        
        // generate a non-repetitive valid coord
        short x, y;
        while (true) {
            uniform_int_distribution<short> uniDist(0, 1000);
            x = uniDist(randomGenerator);
            y = uniDist(randomGenerator);
            
            bool isOk = true;
            for (int u = 0 ; u < cities.size() ; ++u) {
                City &c = cities[u];
                if (abs(x-c.coord.i) < 3 && abs(y-c.coord.j) < 3) {
                    isOk = false;
                    break;
                }
            }
            if (isOk) break;
        }
        coords.emplace_back(x, y);

        int hapiness = getHapiness(level);
        int start, close;
        getStartAndclose(level, start, close);
        int index = cities.size();
        cities.emplace_back(index, name, x, y, hapiness, level, start, close);
        cityNames.push_back(name);
        ++currentCities;
    }
};
void Generator::generateEdgeByCoords() {
    for (short i = 0 ; i < cities.size() - 1 ; ++i) {
        for (short j = i+1 ; j < cities.size() ; ++j) {
            City &c1 = cities[i];
            City &c2 = cities[j];

            int edgeLevel = 4;
            if (c1.level <= 3 && c2.level <= 3) edgeLevel = 3;
            addEdge(i, j, edgeLevel);
        }
    }
};

short Generator::getHapiness(short level) {
    if (level == 1) {
        uniform_int_distribution<short> unidist(100, 150);
        return unidist(randomGenerator);
    }
    if (level == 2) {
        uniform_int_distribution<short> unidist(50, 70);
        return unidist(randomGenerator);
    }
    if (level == 3) {
        uniform_int_distribution<short> unidist(20, 30);
        return unidist(randomGenerator);
    }
    uniform_int_distribution<short> unidist(1, 10);
    return unidist(randomGenerator);

};

void Generator::getStartAndclose(short level, int &start, int &close) {
    if (level == 1) {
        start = 0;
        close = 24 * 60 * 7;
        return;
    }
    if (level == 2) {
        start = 24 * 60 * 1 + 8 * 60; // Mon 0800
        close = 24 * 60 * 5 + 17 * 60; // Friday 1700
        return;
    }

    if (level == 3) {
        uniform_int_distribution<short> unidist(1, 2);
        int d = unidist(randomGenerator);
        start = 24 * 60 * d + 8 * 60; // Monday 8 AM
        close = 24 * 60 * (d+3) + 17 * 60; // Thursday 17:00
        return;
    }

    uniform_int_distribution<short> unidist(1, 5);
    int d = unidist(randomGenerator);
    start = 24 * 60 * (d-1) + 8 * 60;
    close = 24 * 60 * (d+1) + 17 * 60;
}

bool Generator::isDuplicatedEdge(int index1, int index2) {
    for (int i = 0 ; i < edges.size() ; ++i) {
        Edge &e = edges[i];
        if ((index1 == edges[i].city1->index && index2 == edges[i].city2->index)
            || (index1 == edges[i].city2->index && index2 == edges[i].city1->index))  {
            // cout << "Duplicated Edge!" << endl;
            return true;
        }
    }

    // cout << "not duplicated Edge!" << endl;
    return false;
};

int Generator::getTravelTime(short level, double distance) {

    if (level == 1) {
        return distance / 10 + 60;
    }
    if (level == 2) {
        return distance / 6 + 60;
    }
    if (level == 3) {
        return distance / 2;
    }

    return distance;
};

void Generator::saveFile() {
    string filePath = "testCase/C" + to_string(totalCities)+ ".tp.data";
    
    ofstream o(filePath);

    o << totalCities << ' ' << currentEdges << ' ' << timeQuota << ' ' << startTime << endl;
    for (int i = 0 ; i < cities.size() ; ++i) {
        City &c = cities[i];
        o << c.name << ' ' << c.hapiness << ' ' << c.start <<  ' ' << c.close << endl;
    }
    o.flush();
    for (int i = 0 ; i < edges.size() ; ++i) {
        Edge &e = edges[i];
        // o << "edge index: " << i 
        //     << ", city1 index: " << e.city1->index << ", name: " << e.city1->name
        //     << ", city2 index: " << e.city2->index << ", name: " << e.city2->name << endl;
        // o   << "     edge level: " << e.level
        //     << ", distance: " << e.distance
        //     << ", travelTime: " << e.travelTime << endl;
        o << e.city1->name << ' ' << e.city2->name << ' ' << e.travelTime << endl;
        o.flush();
    }

};

void Generator::printTest() {
    /*
    cout << " -----------" << cities.size() << " cities: ----------" << endl;
    cout << totalCities << ' ' << currentEdges << ' ' << timeQuota << ' ' << startTime << endl;
    for (int i = 0 ; i < cities.size() ; ++i) {
        City &c = cities[i];
        cout << c.name << ' ' << c.hapiness << ' ' << c.level << ' ' << c.start <<  ' ' << c.close
            << " | (" << c.coord.i << ", " << c.coord.j << ")" << endl;
    }
    */

    cout << " --------- " << edges.size() << " edges --------------" << endl;

    /*
    for (int i = 0 ; i < edges.size() ; ++i) {
        Edge &e = edges[i];
        cout << "--------" << endl;
        cout << "e.city1->name: " << e.city1.name << endl;
        cout << "e.city2->name: " << e.city2->name << endl;
        cout << e.level << ' ' << e.travelTime << endl;
    }
    */
};

void Generator::printStatus() {
    cout << "cities: " << cities.size() << ", edges: " << edges.size() << endl;
    cout << "currentCities: " << currentCities << ", currentEdges: " << currentEdges<< endl;
};

void addCityAndEdge(Generator &g) {
    g.addCity(1, 270, 136); // 0
    g.addCity(1, 735, 307); // 1
    g.addCity(1, 240, 596); // 2
    g.addCity(1, 829, 804); // 3
    g.addCity(2, 179,  75); // 4

    g.addCity(2, 129, 269); // 5
    g.addCity(2, 136, 538); // 6
    g.addCity(2, 405, 233); // 7
    g.addCity(2, 370, 551); // 8
    g.addCity(2, 613, 146); // 9

    g.addCity(2, 803,  87); // 10
    g.addCity(2, 949, 141); // 11
    g.addCity(2, 852, 399); // 12
    g.addCity(2, 175, 701); // 13
    g.addCity(2, 403, 749); // 14

    g.addCity(2, 685, 825); // 15
    g.addCity(2, 903, 704); // 16
    
    // 6
    g.addEdge(0, 1, 1);
    g.addEdge(0, 2, 1);
    g.addEdge(0, 3, 1);
    g.addEdge(0, 4, 2);
    g.addEdge(0, 5, 2);
    g.addEdge(0, 7, 2);

    // 7
    g.addEdge(1, 0, 1);
    g.addEdge(1, 2, 1);
    g.addEdge(1, 3, 1);
    g.addEdge(1, 9, 2);
    g.addEdge(1, 10, 2);
    g.addEdge(1, 11, 2);
    g.addEdge(1, 12, 2);

    // 7
    g.addEdge(2, 0, 1);
    g.addEdge(2, 1, 1);
    g.addEdge(2, 3, 1);
    g.addEdge(2, 6, 2);
    g.addEdge(2, 8, 2);
    g.addEdge(2, 13, 2);
    g.addEdge(2, 14, 2);

    // 6
    g.addEdge(3, 0, 1);
    g.addEdge(3, 1, 1);
    g.addEdge(3, 2, 1);
    g.addEdge(3, 14, 2);
    g.addEdge(3, 15, 2);
    g.addEdge(3, 16, 2);

    // 3
    g.addEdge(4, 0, 2);
    g.addEdge(4, 5, 2);
    g.addEdge(4, 9, 2);

    // 5
    g.addEdge(5, 0, 2);
    g.addEdge(5, 4, 2);
    g.addEdge(5, 6, 2);
    g.addEdge(5, 7, 2);
    g.addEdge(5, 8, 2);

    // 4
    g.addEdge(6, 2, 2);
    g.addEdge(6, 5, 2);
    g.addEdge(6, 8, 2);
    g.addEdge(6, 13, 2);

    // 4
    g.addEdge(7, 0, 2);
    g.addEdge(7, 5, 2);
    g.addEdge(7, 8, 2);
    g.addEdge(7, 9, 2);

    // 6
    g.addEdge(8, 2, 2);
    g.addEdge(8, 5, 2);
    g.addEdge(8, 6, 2);
    g.addEdge(8, 7, 2);
    g.addEdge(8, 13, 2);
    g.addEdge(8, 14, 2);

    // 4
    g.addEdge(9, 1, 2);
    g.addEdge(9, 4, 2);
    g.addEdge(9, 7, 2);
    g.addEdge(9, 10, 2);

    // 4
    g.addEdge(10, 1, 2);
    g.addEdge(10, 9, 2);
    g.addEdge(10, 11, 2);
    g.addEdge(10, 12, 2);

    // 3
    g.addEdge(11, 1, 2);
    g.addEdge(11, 10, 2);
    g.addEdge(11, 12, 2);
    
    // 5
    g.addEdge(12, 1, 2);
    g.addEdge(12, 10, 2);
    g.addEdge(12, 11, 2);
    g.addEdge(12, 15, 2);
    g.addEdge(12, 16, 2);

    //4
    g.addEdge(13, 2, 2);
    g.addEdge(13, 6, 2);
    g.addEdge(13, 8, 2);
    g.addEdge(13, 14, 2);

    // 5
    g.addEdge(14, 2, 2);
    g.addEdge(14, 3, 2);
    g.addEdge(14, 8, 2);
    g.addEdge(14, 13, 2);
    g.addEdge(14, 15, 2);

    // 4
    g.addEdge(15, 3, 2);
    g.addEdge(15, 12, 2);
    g.addEdge(15, 14, 2);
    g.addEdge(15, 16, 2);

    // 3
    g.addEdge(16, 3, 2);
    g.addEdge(16, 12, 2);
    g.addEdge(16, 15, 2);

    g.printStatus();
};